### PR TITLE
Create m_stun.cpp

### PR DESCRIPTION
**Pull request by ddan39**
Provides the /STUN command, which stops a user from receiving messages/notices to channels

### CODE REVIEW

**Comment by SaberUK** [comment]
This probably should be a user mode rather than an XLine.

File: 2.0/m_stun.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+ public:
+	std::string matchtext;
+
+	Stun(time_t s_time, long d, const std::string& src, const std::string& re, const std::string& stunmask)
+		: XLine(s_time, d, src, re, "STUN")
+		, matchtext(stunmask)
+	{
+	}
+
+	~Stun()
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: You can delete empty blocks like this.

**Comment by ddan39** [comment]
I purposely made an XLine instead of a umode because I do not want the user being notified about it or able to remove it themselves. There is already the +d deaf mode which does that... The use for this module would be for me to use on people who are on my network just to spy on us, and I do not want to provoke them with a zline.

File: 2.0/m_stun.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include "inspircd.h"
+#include "xline.h"
+
+/* $ModDesc: Provides the /STUN command, which stops a user from receiving messages/notices to channels */
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: You are missing a core ModDepends, and ModConfig (see what other modules do).

File: 2.0/m_stun.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+	{
+	}
+
+	bool Matches(User *u)
+	{
+		// E: overrides stun
+		if (u->exempt)
+			return false;
+
+		if (InspIRCd::Match(u->GetFullHost(), matchtext) || InspIRCd::Match(u->GetFullRealHost(), matchtext) || InspIRCd::Match(u->nick+"!"+u->ident+"@"+u->GetIPString(), matchtext))
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: You can just make this into a single return instead of using an if block.

File: 2.0/m_stun.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+			return true;
+
+		return false;
+	}
+
+	bool Matches(const std::string &s)
+	{
+		if (matchtext == s)
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: `return matchtext == s;`

File: 2.0/m_stun.cpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+		return new Stun(set_time, duration, source, reason, xline_specific_mask);
+	}
+
+	bool AutoApplyToUserList(XLine *x)
+	{
+		return false;
+	}
+};
+
+//typedef std::vector<Stun> stunlist;
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: You should probably delete unused code like this.

File: 2.0/m_stun.cpp [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+				user->WriteServ("NOTICE %s :*** Stun for %s already exists", user->nick.c_str(), target.c_str());
+				return CMD_FAILURE;
+			}
+		}
+		return CMD_SUCCESS;
+	}
+
+	RouteDescriptor GetRouting(User* user, const std::vector<std::string>& parameters)
+	{
+		if (IS_LOCAL(user))
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: `return IS_LOCAL(user) ? ROUTE_LOCALONLY : ROUTE_BROADCAST`

File: 2.0/m_stun.cpp [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+	virtual ~ModuleStun()
+	{
+		ServerInstance->XLines->DelAll("STUN");
+		ServerInstance->XLines->UnregisterFactory(&f);
+	}
+
+	/*
+	 * I dont think i need this for OnUserPreNotice and OnUserPreMessage...?
+	 *
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: If you don't need it then delete it.

File: 2.0/m_stun.cpp [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+	void Prioritize()
+	{
+		Module* alias = ServerInstance->Modules->Find("m_alias.so");
+		ServerInstance->Modules->SetPriority(this, I_OnPreCommand, PRIORITY_BEFORE, &alias);
+	}
+	*/
+
+	virtual ModResult OnStats(char symbol, User* user, string_list& out)
+	{
+		if (symbol != 'K')
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: You should change this to something more unique.

File: 2.0/m_stun.cpp [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+		deaf_bypasschars = tag->getString("bypasschars");
+		deaf_bypasschars_uline = tag->getString("bypasscharsuline");
+
+		affectopers = tag->getBool("affectopers", false);
+	}
+
+	/*
+	 * The rest of the code is mostly taken from m_deaf.so
+	 * Copyright (C) 2006-2007 Dennis Friis <peavey@inspircd.org>
```
Comment (SaberUK) [inline 2.0/m_stun.cpp]: If the copyright  is already mentioned in the header you don't need to repeat it.

**Pull request merged**